New package py-pyem@1.2.0

dorton21 opened the pull request.

tgamblin committed:
concretizer: unify logic for spec conditionals

This builds on #20638 by unifying all the places in the concretizer where
things are conditional on specs. Previously, we duplicated a common spec
conditional pattern for dependencies, virtual providers, conflicts, and
externals. That was introduced in #20423 and refined in #20507, and
roughly looked as follows.

Given some directives in a package like:

```python
depends_on("foo@1.0+bar", when="@2.0+variant")
provides("mpi@2:", when="@1.9:")
```

We handled the `@2.0+variant` and `@1.9:` parts by generating generated
`dependency_condition()`, `required_dependency_condition()`, and
`imposed_dependency_condition()` facts to trigger rules like this:

```prolog
dependency_conditions_hold(ID, Parent, Dependency) :-
  attr(Name, Arg1)             : required_dependency_condition(ID, Name, Arg1);
  attr(Name, Arg1, Arg2)       : required_dependency_condition(ID, Name, Arg1, Arg2);
  attr(Name, Arg1, Arg2, Arg3) : required_dependency_condition(ID, Name, Arg1, Arg2, Arg3);
  dependency_condition(ID, Parent, Dependency);
  node(Parent).
```

And we handled `foo@1.0+bar` and `mpi@2:` parts ("imposed constraints")
like this:

```prolog
attr(Name, Arg1, Arg2) :-
  dependency_conditions_hold(ID, Package, Dependency),
  imposed_dependency_condition(ID, Name, Arg1, Arg2).

attr(Name, Arg1, Arg2, Arg3) :-
  dependency_conditions_hold(ID, Package, Dependency),
  imposed_dependency_condition(ID, Name, Arg1, Arg2, Arg3).
```

These rules were repeated with different input predicates for
requirements (e.g., `required_dependency_condition`) and imposed
constraints (e.g., `imposed_dependency_condition`) throughout
`concretize.lp`. In #20638 it got to be a bit confusing, because we used
the same `dependency_condition_holds` predicate to impose constraints on
conditional dependencies and virtual providers. So, even though the
pattern was repeated, some of the conditional rules were conjoined in a
weird way.

Instead of repeating this pattern everywhere, we now have *one* set of
consolidated rules for conditions:

```prolog
condition_holds(ID) :-
  condition(ID);
  attr(Name, A1)         : condition_requirement(ID, Name, A1);
  attr(Name, A1, A2)     : condition_requirement(ID, Name, A1, A2);
  attr(Name, A1, A2, A3) : condition_requirement(ID, Name, A1, A2, A3).

attr(Name, A1)         :- condition_holds(ID), imposed_constraint(ID, Name, A1).
attr(Name, A1, A2)     :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2).
attr(Name, A1, A2, A3) :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2, A3).
```

this allows us to use `condition(ID)` and `condition_holds(ID)` to
encapsulate the conditional logic on specs in all the scenarios where we
need it. Instead of defining predicates for the requirements and imposed
constraints, we generate the condition inputs with generic facts, and
define predicates to associate the condition ID with a particular
scenario. So, now, the generated facts for a condition look like this:

```prolog
condition(121).
condition_requirement(121,"node","cairo").
condition_requirement(121,"variant_value","cairo","fc","True").
imposed_constraint(121,"version_satisfies","fontconfig","2.10.91:").
dependency_condition(121,"cairo","fontconfig").
dependency_type(121,"build").
dependency_type(121,"link").
```

The requirements and imposed constraints are generic, and we associate
them with their meaning via the id. Here, `dependency_condition(121,
"cairo", "fontconfig")` tells us that condition 121 has to do with the
dependency of `cairo` on `fontconfig`, and the conditional dependency
rules just become:

```prolog
dependency_holds(Package, Dependency, Type) :-
  dependency_condition(ID, Package, Dependency),
  dependency_type(ID, Type),
  condition_holds(ID).
```

Dependencies, virtuals, conflicts, and externals all now use similar
patterns, and the logic for generating condition facts is common to all
of them on the python side, as well. The more specific routines like
`package_dependencies_rules` just call `self.condition(...)` to get an id
and generate requirements and imposed constraints, then they generate
their extra facts with the returned id, like this:

```python
    def package_dependencies_rules(self, pkg, tests):
        """Translate 'depends_on' directives into ASP logic."""
        for _, conditions in sorted(pkg.dependencies.items()):
            for cond, dep in sorted(conditions.items()):
                condition_id = self.condition(cond, dep.spec, pkg.name)  # create a condition and get its id
                self.gen.fact(fn.dependency_condition(  # associate specifics about the dependency w/the id
                    condition_id, pkg.name, dep.spec.name
                ))
        # etc.
```

- [x] unify generation and logic for conditions
- [x] use unified logic for dependencies
- [x] use unified logic for virtuals
- [x] use unified logic for conflicts
- [x] use unified logic for externals

LocalWords:  concretizer mpi attr Arg concretize lp cairo fc fontconfig
LocalWords:  virtuals def pkg cond dep fn refactor github py

tgamblin committed:
bugfix: do not generate dep conditions when no dependency

We only consider test dependencies some of the time. Some packages are
*only* test dependencies. Spack's algorithm was previously generating
dependency conditions that could hold, *even* if there was no potential
dependency type.

- [x] change asp.py so that this can't happen -- we now only generate
      dependency types for possible dependencies.

dorton21 committed:
New package py-pyem@1.2.0

Commit:
ascent: Relax CMake version constraints (#22253)

bryanherman committed:
py-h5py: offline installation and deps fix (#22262)

* added h5py patch file for offline installs to work

* h5py v3+ deps consistent with setup.py, add patch

bryanherman committed:
fixed ^python in py-rich, py-dvc (#22264)

vsoch committed:
adding spack -c to set one off config arguments (#22251)

This pull request will add the ability for a user to add a configuration argument on the fly, on the command line, e.g.,:

```bash
$ spack -c config:install_tree:root:/path/to/config.yaml -c packages:all:compiler:[gcc] list --help
```
The above command doesn't do anything (I'm just getting help for list) but you can imagine having another root of packages, and updating it on the fly for a command (something I'd like to do in the near future!)

I've moved the logic for config_add that used to be in spack/cmd/config.py into spack/config.py proper, and now both the main.py (where spack commands live) and spack/cmd/config.py use these functions. I only needed spack config add, so I didn't move the others. We can move the others if there are also needed in multiple places.

sethrj committed:
openblas: prevent microarch flags for generic targets (#22270)

As of OpenBLAS 0.3.13, leaving off `TARGET` by default optimizes most
code for the host system -- adding flags that cause the resulting
library to fail (SIGILL) on older systems.  This change should ensure
that a "x86_64" target for example will work across deployment systems.

https://github.com/xianyi/OpenBLAS/issues/3139

adamjstewart commented on var/spack/repos/builtin/packages/py-pyem/package.py:
```diff
@@ -0,0 +1,33 @@
+    version('2.1.0', sha256='5234a20427ab2813a8a0bf1e9112d4d854b1b0502b3e63d17c1b1a3c4be9340e')
+
+    depends_on('python@3.7:', type=('build', 'run'))
+    depends_on('py-virtualenv', type='build')
+    depends_on('py-pip', type='build')
+    depends_on('py-pytest', type=('build', 'run'))
+    depends_on('py-mypy', type=('build', 'run'))
+    depends_on('py-black', type=('build', 'run'))
+    depends_on('flit', type=('build', 'run'))
```
I don't think any of these deps are needed.

adamjstewart commented on var/spack/repos/builtin/packages/py-pyem/package.py:
```diff
@@ -0,0 +1,33 @@
+    depends_on('py-black', type=('build', 'run'))
+    depends_on('flit', type=('build', 'run'))
+
+    def install(self, spec, prefix):
+        pip = which('pip')
+        pip('install', self.stage.archive_file, '--prefix={0}'.format(prefix))
```
The default install method should work fine, you shouldn't need to override this.

dorton21 replied:
Without overriding the install process none of the pyem commands are recognized. This seemed to be fixed when I added the dep for py-pip and overrode the install phase.

adamjstewart replied:
> none of the pyem commands are recognized

Can you elaborate on this? How can I reproduce this issue?

dorton21 replied:
`spack install py-pyem`
`module load py-pyem`
`module list` *verify the proper module is loaded
`pyem venv add python3.7`
returns "-bash: pyem: command not found"

adamjstewart replied:
Ah, it looks like their build script is broken. The following patch solved the issue for me:
```diff
--- a/setup.py	2021-03-16 16:30:13.000000000 -0500
+++ b/setup.py	2021-03-16 16:30:41.000000000 -0500
@@ -1,7 +1,7 @@
 #!/usr/bin/env python
 # setup.py generated by flit for tools that don't yet use PEP 517
 
-from distutils.core import setup
+from setuptools import setup
 
 packages = \
 ['pyem', 'pyem.projects']
```
You'll also need to add a build/run dep on setuptools. If you have the time, please try to contribute that patch upstream. If not, I can try.

dorton21 replied:
Under what file did you find `a/setup.py	2021-03-16 16:30:13.000000000 -0500` and `from distutils.core import setup`? This is a minor package for me so I'll probably just filter_file it out unless you want to fix it upstream.

adamjstewart replied:
This patch applies to the `setup.py` file.

dorton21 replied:
I just used filter_file and it seems to install and work fine now.

tgamblin committed:
bugfix: allow imposed constraints to be overridden in special cases

In most cases, we want condition_holds(ID) to imply any imposed
constraints associated with the ID. However, the dependency relationship
in Spack is special because it's "extra" conditional -- a dependency
*condition* may hold, but we have decided that externals will not have
dependencies, so we need a way to avoid having imposed constraints appear
for nodes that don't exist.

This introduces a new rule that says that constraints are imposed
*unless* we define `do_not_impose(ID)`. This allows rules like
dependencies, which rely on more than just spec conditions, to cancel
imposed constraints.

We add one special case for this: dependencies of externals.

a-saitoh-fj committed:
py-botocore: Update version to install py-s3fs (#22257)

* Update version to install py-s3fs

* Update the required version of py-urllib3, and Set the dependent conditions of py-docutils

vsoch committed:
do not validate variants of concrete specs in solver setup (#22272)

Currently, regardless of a spec being concrete or not, we validate its variants in `spec_clauses` (part of `SpackSolverSetup`).  

This PR skips the check if the spec is concrete.

The reason we want to do this is so that the solver setup class (really, `spec_clauses`) can be used for cases when we just want the logic statements / facts (is that what they are called?) and we don't need to re-validate an already concrete spec.  We can't change existing concrete specs, and we have to be able to handle them *even if they violate constraints in the current spack*.  This happens in practice if we are doing the validation for a spec produced by a different spack install.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

a-saitoh-fj committed:
py-aioitertools: New package (#22256)

* py-aioitertools: New package

* py-aioitertools: Fixed python dependencies, and Fixed style of code

* switch to the source build

junghans committed:
votca-*: add v2021 (#22282)

t-nojiri committed:
fortrilinos: Fix trilinos depend (#22295)

adamjstewart committed:
PyTorch: add missing tqdm dependency (#22289)

adamjstewart committed:
py-tqdm: add new version, variants (#22290)

adamjstewart committed:
Fix use of quotes in Python build system (#22279)

adamjstewart committed:
dos2unix: gettext needed for linking (#22288)

haampie committed:
Make py-boto3 installable again with old concretizer (#22298)

dorton21 committed:
Changed deps

Commit:
vtk-h: Fix missing trailing-colon in CMake version (#22302)

haampie committed:
Propagate --test= for environments (#22040)

* Propagate --test= for environments

* Improve help comment for spack concretize --test flag

* Add tests for --test with environments

haampie committed:
Expand relative dev paths in environment files (#22045)

* Rewrite relative dev_spec paths internally to absolute paths in case of relocation of the environment file

* Test relative paths for dev_path in environments

* Add a --keep-relative flag to spack env create

This ensures that relative paths of develop paths are not expanded to
absolute paths when initializing the environment in a different location
from the spack.yaml init file.

adamjstewart committed:
OpenCV: add new version, simplify package (#22292)

haampie committed:
hwloc: add dependency on ncurses (#22300)

Co-authored-by: Michal Sudwoj <msudwoj@student.ethz.ch>

t-nojiri committed:
libflame: Fix build error (#22315)

t-nojiri committed:
amdlibflame: Fix build error (#22316)

a-saitoh-fj committed:
py-aiobotocore: New package (#22258)

* py-aiobotocore: New package

* py-aiobotocore: Added python dependencies, and Removed unnecessary whitespace

alalazo committed:
Speed-up CI by reorganizing tests (#22247)

* unit tests: mark slow tests as "maybeslow"

This commit also removes the "network" marker and
marks every "network" test as "maybeslow". Tests
marked as db are maintained, but they're not slow
anymore.

* GA: require style tests to pass before running unit-tests

* GA: make MacOS unit tests fail fast

* GA: move all unit tests into the same workflow, run style tests as a prerequisite

All the unit tests have been moved into the same workflow so that a single
run of the dorny/paths-filter action can be used to ask for coverage based
on the files that have been changed in a PR. The basic idea is that for PRs
that introduce only changes to packages coverage is not necessary, this
resulting in a faster execution of the tests.

Also, for package only PRs slow unit tests are skipped.

Finally, MacOS and linux unit tests are now conditional on style tests passing
meaning that e.g. we won't waste a MacOS worker if we know that the PR has
flake8 issues.

* Addressed review comments

* Skipping slow tests on MacOS for package only recipes

* QA: make tests on changes correct before merging

alalazo committed:
The action to detect changes needs a repository to be checked out on push event (#22324)

amd-toolchain-support committed:
WRF package: add aocc 3.0 support for WRF 3.9.1.1 and 4.2 (#22285)

* Added 2 new configure patch files to build WRF 3.9.1.1 and 4.2
  with aocc@3.0
* Renamed patch files used for building WRF 3.9.1.1 and 4.2 with
  aocc@2.3 (mostly, this also removes -march=native from AOCCOPT
  and updates LIBMVEC options for aocc@2.3)

wdconinc committed:
containerize: fix typo in documentation (#22331)

Before this fix, `spack containerize` complains that `centos/7` is invalid
(should have been `centos:7`)

amd-toolchain-support committed:
spack recipe for aocl-sparse (#22079)

cosmicexplorer committed:
fix weird failure in variant values (#22328)

dev-zero committed:
llvm: enable ncurses patch for building on SUSE in 11.0.1 (#22319)

fixes #19625 again

eugeneswalker committed:
amrex: add gcc 8 conflicts (#22339)

t-nojiri committed:
r-v8: Fix build error (#22314)

bryanherman committed:
add maintainer to py-h5py (#22332)

davehadley committed:
pythia6: fix wrong patch (#22338)

See #22330

eschnett committed:
ssht: add v1.3.5 (#22306)

btovar committed:
cctools: add v7.2.2 (#22325)

haampie committed:
boost: fix unicode bugs in Boost recipe (#22301)

qwertos committed:
flac: add v1.3.3 (#22307)

eschnett committed:
fasttransforms: add v0.5.0 (#22304)

eschnett committed:
shtools: add v4.8 (#22305)

haampie committed:
spack location: bugfix for out of source build dirs (#22348)

amd-toolchain-support committed:
aocc: add support for v3.0 compilers (#22219)

A mitigation of a known issue that affects v3.0 is added, see
https://developer.amd.com/wp-content/resources/AOCC-3.0-Install-Guide.pdf

ChristianTackeGSI committed:
root: add v6.22.08 (#22263)

adamjstewart committed:
tidy-html5: add new package (#22286)

adamjstewart committed:
GDAL: add v3.2.2 (#22224)

t-karatsu committed:
intel-mpi-benchmarks: Take in upstream bugfix patches. (#22317)

haampie committed:
Fix indentation compiler wrapper issue (#22352)

sethrj committed:
openblas: fix older versions (#22358)

hseara committed:
Amber (#22275)

* Improved amber package. Added amber20

* Corrected description link between ambertools and amber versions

* Fixed ambertools handling

* Apply pathes only when problematic traget requested

* fixed style issue

* Added amber20 sha256 checksum

* Improved documentation

* Removed preferred amber18

* Changef f-string to str.format()

* Changef f-string to str.format()

* removed url_for_version fuction

haampie committed:
Tab to spaces (#22362)

btovar committed:
CCTools: add v7.2.3 (#22355)

alalazo committed:
archspec: update to latest version (#22357)

ax3l committed:
py-yt: 3.6.1 (#22329)

* py-yt: 3.6.1

The latest bugfix release of yt-project fixes issues with matplotlib
version 3.3.0+.

* py-yt: add many more version constrains

rhoneyager committed:
Add udunits2 v2.2.28 and target the new UCAR repository (#22370)

haampie committed:
Fix broken spack -c flag (#22361)

kuberry committed:
xyce: add new package (#22345)

cessenat committed:
perl-tk: creating package (#21586)

dorton21 committed:
Fixes to install of pyem

dorton21 committed:
Merge issues

alalazo commented:
I think there have been merge issues in this PR

dorton21 commented:
> I think there have been merge issues in this PR

Should I close this branch and resubmit it. I figure this will fix the merge issues.
